### PR TITLE
Cypress article

### PR DESCRIPTION
**Pull request by sapegin**

### CODE REVIEW

File: content/all/react-testing-4-cypress.md [hunk 1]
```diff
@@ -15,40 +15,42 @@ tags:
   - msw
 ---
 
-Cypress is a framework-agnostic end-to-end (E2E) testing tool for web apps. Together with Cypress Testing Library, it gives the best test writing experience, and makes applying best practices, we’ve learned in the first article, easy.
+Cypress is a framework-agnostic end-to-end testing (also known as E2E, or integration testing) tool for web apps. Together with Cypress Testing Library and Mock Service Worker, it gives the best test writing experience and makes writing good, resilient to changes, tests straightforward.
 
-**This is the fourth article in the series**, where we learn how to test React apps with Cypress and Cypress Testing Library.
+**This is the fourth article in the series**, where we learn how to test React apps end-to-end using Cypress and Cypress Testing Library, and how to mock network requests using Mock Service Worker.
```
Comment (crecotun) [inline content/all/react-testing-4-cypress.md:20]: Sounds a bit unnatural to me.
Would it make sense to move **end-to-end** before **React apps**?

File: content/all/react-testing-4-cypress.md [hunk 2]
```diff
@@ -15,40 +15,42 @@ tags:
...
 - Supports Chrome, Firefox and Edge.
 
 **Cypress Testing Library** makes Cypress even better:
 
-- Convenient queries, like finding elements by their label text or ARIA role.
+- Convenient semantic queries, like finding elements by their label text or ARIA role.
 - Libraries for other frameworks with the same queries.
 
-Testing Library helps you write good tests and makes writing bad tests difficult. It allows you to query elements similar to how a user would do that: for example, query form elements and buttons by their labels.
+Testing Library helps us write [good tests](/all/react-testing-1-best-practices/) and makes writing bad tests difficult. It allows us to interact with the app similar to how a real user would do that: for example, find form elements and buttons by their labels. It helps us to avoid testing implementation details, making our tests resilient to code changes that don’t change the behavior.
```
Comment (crecotun) [inline content/all/react-testing-4-cypress.md:53]: and makes it difficult to write bad tests?

File: content/all/react-testing-4-cypress.md [hunk 3]
```diff
@@ -103,11 +128,11 @@ import '@testing-library/cypress/add-commands';
 
 ### Setting up Mock Service Worker
 
-We’re going to use [Mock Service Worker](https://mswjs.io/) (MSW) for mocking network requests in our integration tests, and in the app during development. Cypress has [its own way of mocking network](https://docs.cypress.io/api/commands/server.html) but I think MSW has several benefits:
+We’re going to use [Mock Service Worker](https://mswjs.io/) (MSW) for mocking network requests in our integration tests, and in the app during development. Cypress has [its way of mocking network](https://docs.cypress.io/api/commands/server.html), but I think MSW has several benefits:
 
-- Using Service Worker, so it intercepts all network requests, no matter how there are made.
-- A single place do define mocks for the project, with the ability to [override responses](https://mswjs.io/docs/api/setup-server/use) for particular tests.
-- We can reuse mocks in integration tests and during development.
+- Uses Service Workers, so it intercepts all network requests, no matter how there are made.
```
Comment (crecotun) [inline content/all/react-testing-4-cypress.md:133]: It uses?

File: content/all/react-testing-4-cypress.md [hunk 4]
```diff
@@ -103,11 +128,11 @@ import '@testing-library/cypress/add-commands';
...
+- A single place to define mocks for the project, with the ability to [override responses](https://mswjs.io/docs/api/setup-server/use) for particular tests.
+- Ability to reuse mocks in integration tests and during development.
```
Comment (crecotun) [inline content/all/react-testing-4-cypress.md:135]: An ability?

File: content/all/react-testing-4-cypress.md [hunk 5]
```diff
@@ -490,17 +563,25 @@ With the Cypress’ [viewport()](https://docs.cypress.io/api/commands/viewport.h
 });
 ```
 
-### Debugging
+## Debugging
 
-Cypress docs has a thorough [debugging guide](https://docs.cypress.io/guides/guides/debugging.html).
+Cypress docs have a thorough [debugging guide](https://docs.cypress.io/guides/guides/debugging.html).
```
Comment (crecotun) [inline content/all/react-testing-4-cypress.md:568]: I think **has** is better here as you refer to Cypress Documentation 🤔

Comment (sapegin) [reply content/all/react-testing-4-cypress.md:568]: Grammarly may be confused by the word “docs”, I'm not sure what's correct.

File: content/all/react-testing-4-cypress.md (hunk 1 above)
Comment (sapegin) [reply content/all/react-testing-4-cypress.md:20]: I don't feel that either of the two is more natural ;-/

File: content/all/react-testing-4-cypress.md (hunk 2 above)
Comment (sapegin) [reply content/all/react-testing-4-cypress.md:53]: Why do you prefer a more complex variant?

**Pull request merged**